### ISSUE TITLE
Unable to identify Pop-up dialog and its UI elements from Inspect.exe

### ISSUE THREAD

**Issue by ManasaBhat1** [microsoft/WinAppDriver#1959]
Unable to identify Pop-up dialog and its UI elements from Inspect.exe. It was all working fine before, after moving everything to a new process, its not identifying the UI elements of a pop-up dialogs.  

Before only one 32 bit application was there now main GUI and its functionalities[code] moved to a x64 bit application. 32 bit application host this x64 bit application. 

Please let me know any solution which could help our investigation. its important for us to identify elements in Inspect tool as same way TFA the automation scripts are done.

**Comment by RakeshDangi-slt** [comment]
Hi @ManasaBhat1, Try seeing the popup from the root level, It might have moved to the root level during this change, you wont find it in the main application therefore switch to root and then try to find the element.

**Comment by ManasaBhat1** [comment]
Thank you @RakeshDangi-slt for the response. What we have identified was for these dialogs the "Owner" is not set after the architectural change. We are manually setting the root as their owner and seems to be working. But it has to be manually done wherever this "Showdailog" function is getting called. 

It may not be the right forum to ask this , but still would really appreciate any help. Is there any generic way to do this instead of making the change everywhere manually? Btw, the root element for all the dialog form is same.

**Comment by RakeshDangi-slt** [comment]
Yes you have to launch the win app driver at the root level, launch your application seprately and then you can switch between windows from root level easily.
You can switch between windows using native window handle like this: try {
            // Retrieve the target element and its native window handle
            WebElement targetElement = utilFunc.getWindowsElement(elementName, 30);
            String nativeWindowHandle = targetElement.getAttribute("NativeWindowHandle");
         // Convert nativeWindowHandle to the desired format
            nativeWindowHandle = "0x" + Integer.toHexString(Integer.parseInt(nativeWindowHandle));
            switchFocusToWindowHandle(nativeWindowHandle);
        } catch (Exception e) {
            String errorMessage = "Error switching focus to element: " + elementName + " due to: " + e.getMessage();
            log.error(errorMessage, e);
            throw new ErrorMessage(errorMessage);
        }
    }
    
and :   
    public void switchFocusToWindowHandle(String nativeWindowHandle) throws ErrorMessage {
        try {
            // Create a WindowsDriver with the obtained nativeWindowHandle to switch focus
            DesiredCapabilities updatedCapabilities = buildDesiredCapabilities("appTopLevelWindow", nativeWindowHandle);
            createWindowsDriver(updatedCapabilities);
        } catch (Exception e) {
            String errorMessage = "Error switching focus to window handle: " + nativeWindowHandle + " due to: " + e.getMessage();
            log.error(errorMessage, e);
            throw new ErrorMessage(errorMessage);
        }
    }
    If you need anything else do let me know.
    also may I know the purpose of your development so I can guide better.

**Comment by ManasaBhat1** [comment]
We use TAF (Test Automation Framework} to develop automation scripts, and with the help of inspect.exe to identify the tree of each UI objects. 
With your given solution, it looks like we may need to implement timer concept to identify whenever the dialog [child window] is launched coz we have multiple such dialogs.

**Comment by RakeshDangi-slt** [comment]
We have developed a software for the same purpose you can check it out over here 'https://soliterata.com/', yes you need to implement a time, you can just try to search for it multiple times until a specified time out is reached or you can use appium combined with win app driver like this:public RemoteWebElement getWindowsElement(String variableName, Integer timeoutSeconds) throws ErrorMessage {
    	EnvironmentObjDTO objDTO = this.checkElementIsAvailableOnFile(variableName);

	    if (timeoutSeconds == null) {
	        Object timeoutValue = new DesktopParamClass().getExecutionParamValue(DesktopParamKeyName.WAIT_UNTIL_SECONDS);
	        if (timeoutValue instanceof String) {
	            timeoutSeconds = Integer.parseInt((String) timeoutValue);
	        } else if (timeoutValue instanceof Integer) {
	            timeoutSeconds = (Integer) timeoutValue;
	        } else {
	            timeoutSeconds = Integer.parseInt(DesktopParamKeyName.WAIT_UNTIL_SECONDS); // Defaulting to the static value if the retrieved value is not as expected.
	        }
	    }
	    
	    RemoteWebElement element = null;
        By elemtntRecivedFromBy = null;
        WindowsDriver windowsDriver = (WindowsDriver)(new DesktopParamClass().getExecutionParamValue(DesktopParamKeyName.WIN_APP_DRIVER));
//        windowsDriver.getPageSource();
        
        //System.out.println("page ssssource:"+ windowsDriver.getPageSource());
        WebDriverWait initialWait = new WebDriverWait(windowsDriver, Duration.ofSeconds(timeoutSeconds / 2));
	    // Full timeout for extended wait
	    WebDriverWait extendedWait = new WebDriverWait(windowsDriver, Duration.ofSeconds(timeoutSeconds / 2));
       
        try {
            elemtntRecivedFromBy = find(objDTO.getElementType(), objDTO.getElementValue());
            element = (RemoteWebElement)initialWait.until(ExpectedConditions.visibilityOfElementLocated(elemtntRecivedFromBy));
//            List<WebElement> elements = windowsDriver.findElements(elemtntRecivedFromBy);
//            System.out.println("elements:"+ elements);
            //this.switchToElementWithNativeWindowHandle(element,elemtntRecivedFromBy);
            log.info("The element is found with value: " + element);

            // highlighting it for screenshot
            this.highlightElement(element);
        } catch (TimeoutException e) {
            log.info("First instance failed, element is not visible, trying to find in DOM");
            element = (RemoteWebElement)extendedWait.until(ExpectedConditions.presenceOfElementLocated(find(objDTO.getElementType(), objDTO.getElementValue())));
            //this.switchToElementWithNativeWindowHandle(element,elemtntRecivedFromBy);

            // highlighting it for screenshot
            this.highlightElement(element);
        } catch (Exception e) {
            log.error("Timeout occurred before element is visible: " + e.getMessage());
            throw new TimeoutException("Timeout occurred before element was visible. Timeout is of:" + timeoutSeconds);
        }
        return element;
    }

**Comment by ManasaBhat1** [comment]
Your solution is good, but doesn't seems to be working for us. Child window gets called before its fully loaded and initialized If we try to set it from the root class with timer.  Trying to with some of its form properties to identify whether its already been loaded or not before setting the owner for it.  we are still working on it.  the child windows are not direct, its very nested, thats the reason we are not able to set owner directly without impacting the shared components where these forms dialogs are present.

**Comment by RakeshDangi-slt** [comment]
I am not able to understand properly can you please share some SS of the tree view and others if possible.
